yield_plugin/src/yield_plugin.cpp

```diff
@@ -403,106 +415,243 @@ namespace yield_plugin
+
+        if (i ==0 && j ==0 && distance > config_.collision_check_radius)
+        {
+          RCLCPP_DEBUG_STREAM(rclcpp::get_logger("yield_plugin"), "Too far away" );
+          return boost::none;
+        }
+
+        if (distance > collision_radius) 
+        {
+          continue;
+        }
+
+        // if within collision radius
+        lanelet::BasicPoint2d vehicle_point(x1,y1);
+        lanelet::BasicPoint2d object_point(x2,y2);
+        double vehicle_downtrack = wm_->routeTrackPos(vehicle_point).downtrack;
+        double object_downtrack = wm_->routeTrackPos(object_point).downtrack;
+
+        if (vehicle_downtrack > object_downtrack + config_.vehicle_length / 2)  // if half a length of the vehicle past, it is considered behind
+        {
+          RCLCPP_INFO_STREAM(rclcpp::get_logger("yield_plugin"), "Detected an object nearby behind the vehicle at timestamp " << std::to_string(p2a_t));
+          return boost::none;
+        }
+        else
+        {
+          RCLCPP_WARN_STREAM(rclcpp::get_logger("yield_plugin"), "Collision detected at timestamp " << std::to_string(p2a_t) << ", x: " << x1 << ", y: " << y1 << ", within distance: " << distance);
+          return rclcpp::Time(p2a.header.stamp);
+        }      
+      }
+    }
+
+    // No collision detected
+    return boost::none;
+  }
+
+  carma_planning_msgs::msg::TrajectoryPlan YieldPlugin::update_traj_for_object(const carma_planning_msgs::msg::TrajectoryPlan& original_tp, double initial_velocity) 
+  {    
+    if (original_tp.trajectory_points.empty())
+    {
+      RCLCPP_ERROR_STREAM(rclcpp::get_logger("yield_plugin"), "Yield plugin received empty trajectory plan in update_traj_for_object");
+      throw std::invalid_argument("Yield plugin received empty trajectory plan in update_traj_for_object");
+    }
+
+    rclcpp::Time plan_start_time = original_tp.trajectory_points[0].target_time;
     carma_planning_msgs::msg::TrajectoryPlan update_tpp_vector;
     geometry_msgs::msg::Twist current_velocity;
     current_velocity.linear.x = initial_velocity;
 
-    std::vector<carma_perception_msgs::msg::RoadwayObstacle> rwol = wm_->getRoadwayObjects();
-    carma_perception_msgs::msg::RoadwayObstacleList rwol2;
-    rwol2.roadway_obstacles = rwol;
-    host_vehicle_size.x = config_.vehicle_length;
-    host_vehicle_size.y = config_.vehicle_width;
-    host_vehicle_size.z = config_.vehicle_height; 
+    boost::optional<carma_perception_msgs::msg::ExternalObject> earliest_collision_obj = boost::none;
+    std::set<int> checked_external_object_ids;
 
+    std::vector<carma_perception_msgs::msg::PredictedState> new_list;
+    RCLCPP_DEBUG_STREAM(rclcpp::get_logger("yield_plugin"), "ExternalObjects size: " << external_objects_.size());
 
-    std::vector<carma_perception_msgs::msg::RoadwayObstacle> rwol_collision;
+    if (!wm_->getRoute())
```

Comment:
Replacing this with `wm_->getRoute() != nullptr` reduces the mental overhead in my opinion. In its form now, users have to look up `getRoute` to see what it returns. Is it a boolean, `std::optional`, etc.